lib/utils/log.go

```diff
@@ -21,53 +24,100 @@ const (
+}
+
+var tagToString = map[int]string{
+	ErrorTAG: "[ERROR]",
+	InfoTAG:  "[INFO]",
+	DebugTAG: "[DEBUG]",
+}
+
+var (
+	logfile *os.File
+	mutex   = sync.Mutex{}
+)
+
+func getTimeStamp() string {
+	return fmt.Sprintf("%v", time.Now().Unix())
+}
+
+func coloredContext(context string) string {
+	return magenta + "(" + reset + cyan + context + reset + magenta + ")" + reset
+}
+
+func out(context, message string, TAG int) {
 	currentTime := time.Now()
+	hours := fmt.Sprintf("%d", currentTime.Hour())
+	if currentTime.Hour() < 10 {
+		hours = "0" + hours
+	}
+	minutes := fmt.Sprintf("%d", currentTime.Minute())
+	if currentTime.Minute() < 10 {
+		minutes = "0" + minutes
+	}
+	seconds := fmt.Sprintf("%d", currentTime.Second())
+	if currentTime.Second() < 10 {
+		seconds = "0" + seconds
+	}
 	timeLog := fmt.Sprintf(
-		"%d-%d-%d %d:%d:%d",
+		"%d-%d-%d %s:%s:%s",
 		currentTime.Day(),
 		currentTime.Month(),
 		currentTime.Year(),
-		currentTime.Hour(),
-		currentTime.Minute(),
-		currentTime.Second(),
+		hours,
+		minutes,
+		seconds,
 	)
-	fmt.Println(TAG + reset + " " + yellow + timeLog + reset + lightRed + " >>> " + reset + green + s + reset)
+	fmt.Println(tagToStringColored[TAG] + " " + coloredContext(context) + " " + yellow + timeLog + reset + lightRed + " >>> " + reset + green + message + reset)
+	go func() {
+		mutex.Lock()
+		defer mutex.Unlock()
+		logfile.WriteString(fmt.Sprintf("%s (%s) %s >>> %s\n", tagToString[TAG], context, timeLog, message))
+	}()
 }
 
 // Log logs to the console with your custom TAG
-func Log(s, TAG string) {
-	out(s, TAG)
+func Log(context, message string, TAG int) {
+	out(context, message, TAG)
 }
 
 // LogInfo logs information to the console
-func LogInfo(f string, v ...interface{}) {
-	s := fmt.Sprintf(f, v...)
-	out(s, InfoTAG)
+func LogInfo(context, template string, args ...interface{}) {
+	out(context, fmt.Sprintf(template, args...), InfoTAG)
 }
 
 // LogDebug logs debug messages to console
-func LogDebug(f string, v ...interface{}) {
-	s := fmt.Sprintf(f, v...)
-	out(s, DebugTAG)
+func LogDebug(context, template string, args ...interface{}) {
+	out(context, fmt.Sprintf(template, args...), DebugTAG)
 }
 
 // LogError logs type error to console
-func LogError(e error) {
-	if e == nil {
+func LogError(context string, err error) {
+	if err == nil {
 		return
 	}
-	s := e.Error()
-	out(s, ErrorTAG)
+	out(context, err.Error(), ErrorTAG)
 }
 
 // LogResErr logs type ResponseError to console
-func LogResErr(e types.ResponseError) {
-	s := fmt.Sprintf("%d: %s\n%s", e.Status(), e.Message(), e.Verbose())
-	out(s, ErrorTAG)
+func LogResErr(context string, e types.ResponseError) {
+	message := fmt.Sprintf("%d: %s\n%s", e.Status(), e.Message(), e.Verbose())
+	out(context, message, ErrorTAG)
+}
+
+func init() {
+	if configs.GasperConfig.Debug {
+		logfile, _ = os.Create("gasper.log")
+	} else {
+		logfile, _ = os.Create("gasper-" + getTimeStamp() + ".log")
```

Comment:
Isn't `timestamp` a redundant info for log file titles as they would already contain the timestamps.

Reply:
Oh alright, it registers it whenever gasper starts. Makes sense then.